### PR TITLE
Initial Port of Fabric Biome API

### PR DESCRIPTION
**Pull request by peaceheis**
This is a 1-to-1 port of Fabric's Biome API - I intend on looking at improving the API, but this is just an initial draft to get the ball rolling so the rest of the QSL Worldgen Team has access. 

~~There is one main issue that I have found - the implementation of Biome effects modification is very much broken - looking testing with the testmod and comparing with what the code intends is the best way to see the problem. While feature modification applies correctly, effects like sky color and precipitation do not get applied. It is important to note that the Fabric API has the same issue, and I intend on creating an issue for their sake as well.~~ This is an issue with TestMod. 

~~I intended on fixing this bug myself, but I'm having a hard time figuring out where exactly the issue lies, and hence I'm creating this PR before the API is in a finished state.~~ The so-called "bug" has been isolated and fixed.

### CODE REVIEW

File: library/worldgen/dimension/build.gradle [hunk 1]
```diff
@@ -8,6 +8,7 @@ qslModule {
 	id = "quilt_dimension"
 	library = "worldgen"
 	description = "An API for managing custom dimensions."
+	version = "1.0.0"
```
Comment (LambdAurora) [inline library/worldgen/dimension/build.gradle]: This must be removed.

**Comment by zOnlyKroks** [comment]
The AccessWidener somehow failes. Not sure why yet. Have tinkered around for a bit, mixin accessors also dont seem to work for some reason.

**Comment by LambdAurora** [comment]
> The AccessWidener somehow failes. Not sure why yet. Have tinkered around for a bit, mixin accessors also dont seem to work for some reason.

For the access widener, please look into the buildscript of other modules, it has changed.

**Comment by peaceheis** [comment]
Thanks to @zOnlyKroks, we have determined the issue is actually an issue with Test Mod, and thus Biome API isn't broken, huzzah! File reading may have an issue, though if tags are already created by mod authors in code I can't see this being an issue.

File: library/worldgen/biome/src/main/java/org/quiltmc/qsl/worldgen/biome/impl/modification/BiomeModificationContextImpl.java [hunk 2]
```diff
@@ -0,0 +1,406 @@
...
+import net.minecraft.world.gen.GenerationStep;
+import net.minecraft.world.gen.carver.ConfiguredCarver;
+import net.minecraft.world.gen.feature.ConfiguredStructureFeature;
+import net.minecraft.world.gen.feature.Feature;
+import net.minecraft.world.gen.feature.PlacedFeature;
+import org.jetbrains.annotations.ApiStatus;
+import org.jetbrains.annotations.NotNull;
+import org.quiltmc.qsl.worldgen.biome.api.BiomeModificationContext;
+
+import java.util.*;
```
Comment (NoahvdAa) [inline library/worldgen/biome/src/main/java/org/quiltmc/qsl/worldgen/biome/impl/modification/BiomeModificationContextImpl.java]: wildcard import

File: library/worldgen/biome/src/main/java/org/quiltmc/qsl/worldgen/biome/impl/modification/BiomeModificationImpl.java [hunk 3]
```diff
@@ -0,0 +1,228 @@
...
+import net.minecraft.world.level.LevelProperties;
+import org.jetbrains.annotations.ApiStatus;
+import org.jetbrains.annotations.TestOnly;
+import org.quiltmc.qsl.worldgen.biome.api.BiomeModificationContext;
+import org.quiltmc.qsl.worldgen.biome.api.BiomeSelectionContext;
+import org.quiltmc.qsl.worldgen.biome.api.ModificationPhase;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.util.*;
```
Comment (NoahvdAa) [inline library/worldgen/biome/src/main/java/org/quiltmc/qsl/worldgen/biome/impl/modification/BiomeModificationImpl.java]: same here

File: library/worldgen/biome/src/main/java/org/quiltmc/qsl/worldgen/biome/mixin/MixinMultiNoiseBiomeSource.java [hunk 4]
```diff
@@ -0,0 +1,54 @@
...
+	private static void appendNetherBiomes(Registry<Biome> registry, CallbackInfoReturnable<MultiNoiseUtil.ParameterRangeList<Holder<Biome>>> cri) {
+		MultiNoiseUtil.ParameterRangeList<Holder<Biome>> biomes = cri.getReturnValue();
+		List<Pair<MultiNoiseUtil.NoiseHypercube, Holder<Biome>>> entries = new ArrayList<>(biomes.getEntries());
+
+		// add fabric biome noise point data to list && BiomeSource biome list
+		NetherBiomeData.getNetherBiomeNoisePoints().forEach((biomeKey, noisePoint) -> {
+			entries.add(Pair.of(noisePoint, registry.getHolderOrThrow(biomeKey)));
+		});
+
+		cri.setReturnValue(new MultiNoiseUtil.ParameterRangeList<>(entries));
```
Comment (NoahvdAa) [inline library/worldgen/biome/src/main/java/org/quiltmc/qsl/worldgen/biome/mixin/MixinMultiNoiseBiomeSource.java:51]: might as well overwrite if we unconditionally set the return value? Not sure what our 'policy' on this is.

File: library/worldgen/biome/src/main/resources/fabric.mod.json [hunk 5]
```diff
@@ -0,0 +1,43 @@
+{
```
Comment (NoahvdAa) [inline library/worldgen/biome/src/main/resources/fabric.mod.json]: This file is automatically generated by the gradle task.

File: library/worldgen/biome/src/testmod/java/org/quiltmc/qsl/worldgen/biome/QuiltBiomeTest.java [hunk 6]
```diff
@@ -0,0 +1,128 @@
...
+import net.minecraft.world.biome.*;
+import net.minecraft.world.biome.source.util.MultiNoiseUtil;
+import net.minecraft.world.gen.GenerationStep;
+import net.minecraft.world.gen.decorator.BiomePlacementModifier;
+import net.minecraft.world.gen.decorator.InSquarePlacementModifier;
+import net.minecraft.world.gen.feature.*;
+import net.minecraft.world.gen.feature.util.PlacedFeatureUtil;
+import org.quiltmc.loader.api.ModContainer;
+import org.quiltmc.qsl.base.api.entrypoint.ModInitializer;
+import org.quiltmc.qsl.worldgen.biome.api.*;
```
Comment (NoahvdAa) [inline library/worldgen/biome/src/testmod/java/org/quiltmc/qsl/worldgen/biome/QuiltBiomeTest.java]: evil wildcard impots, although it should matter less here

**Review by NoahvdAa — APPROVED** [review]
Looks good from web apart from the comments below!

File: library/worldgen/biome/src/main/java/org/quiltmc/qsl/worldgen/biome/impl/modification/BiomeModificationContextImpl.java (hunk 2 above)
Comment (zOnlyKroks) [reply library/worldgen/biome/src/main/java/org/quiltmc/qsl/worldgen/biome/impl/modification/BiomeModificationContextImpl.java]: done

File: library/worldgen/biome/src/main/java/org/quiltmc/qsl/worldgen/biome/impl/modification/BiomeModificationImpl.java (hunk 3 above)
Comment (zOnlyKroks) [reply library/worldgen/biome/src/main/java/org/quiltmc/qsl/worldgen/biome/impl/modification/BiomeModificationImpl.java]: done

File: library/worldgen/biome/src/testmod/java/org/quiltmc/qsl/worldgen/biome/QuiltBiomeTest.java (hunk 6 above)
Comment (zOnlyKroks) [reply library/worldgen/biome/src/testmod/java/org/quiltmc/qsl/worldgen/biome/QuiltBiomeTest.java]: done

File: library/worldgen/biome/src/main/java/org/quiltmc/qsl/worldgen/biome/mixin/MixinMultiNoiseBiomeSource.java (hunk 4 above)
Comment (zOnlyKroks) [reply library/worldgen/biome/src/main/java/org/quiltmc/qsl/worldgen/biome/mixin/MixinMultiNoiseBiomeSource.java:51]: done

**Review by zOnlyKroks — APPROVED** [review]

**Review by oliviathevampire — APPROVED** [review]

**Comment by zOnlyKroks** [comment]
@LambdAurora Ur review is needed please :D (I think)

File: library/worldgen/biome/src/main/java/org/quiltmc/qsl/worldgen/biome/api/BiomeSelectionContext.java [hunk 7]
```diff
@@ -0,0 +1,154 @@
...
+	 * used by the current world-save.
+	 *
+	 * <p>If no dimension options exist for the given dimension key, <code>false</code> is returned.
+	 */
+	boolean canGenerateIn(RegistryKey<DimensionOptions> dimensionKey);
+
+	/**
+	 * {@return true if this biome is in the given {@link TagKey}}.
+	 */
+	boolean hasTag(TagKey<Biome> tag);
```
Comment (OroArmor) [inline library/worldgen/biome/src/main/java/org/quiltmc/qsl/worldgen/biome/api/BiomeSelectionContext.java]: i would use `isIn`, which is the standard for tag checking

File: library/worldgen/biome/src/main/java/org/quiltmc/qsl/worldgen/biome/api/BiomeSelectors.java [hunk 8]
```diff
@@ -0,0 +1,182 @@
...
+	public static Predicate<BiomeSelectionContext> foundInTheEnd() {
+		return context -> context.canGenerateIn(DimensionOptions.END);
+	}
+
+	/**
+	 * Returns a biome selector that will match all biomes in the given tag.
+	 *
+	 * @see net.minecraft.tag.TagKey
+	 */
+	public static Predicate<BiomeSelectionContext> tag(TagKey<Biome> tag) {
```
Comment (OroArmor) [inline library/worldgen/biome/src/main/java/org/quiltmc/qsl/worldgen/biome/api/BiomeSelectors.java]: `isIn` again

**Review by OroArmor — COMMENTED** [review]
Small nitpicks, not really an impl review since its a port

File: library/worldgen/biome/src/main/java/org/quiltmc/qsl/worldgen/biome/api/BiomeSelectors.java (hunk 8 above)
Comment (peaceheis) [reply library/worldgen/biome/src/main/java/org/quiltmc/qsl/worldgen/biome/api/BiomeSelectors.java]: Will do

File: library/worldgen/biome/src/main/java/org/quiltmc/qsl/worldgen/biome/impl/modification/BiomeModificationMarker.java [hunk 9]
```diff
@@ -0,0 +1,29 @@
...
+
+import org.jetbrains.annotations.ApiStatus;
+
+/**
+ * Prevents double-modification of biomes in the same dynamic registry manager from occurring and fails-fast
+ * if it does occur.
+ */
+@ApiStatus.Internal
+public interface BiomeModificationMarker {
+	void quilt_markModified();
```
Comment (LambdAurora) [inline library/worldgen/biome/src/main/java/org/quiltmc/qsl/worldgen/biome/impl/modification/BiomeModificationMarker.java]:
```suggestion
	void quilt$markModified();
```

File: library/worldgen/biome/src/main/java/org/quiltmc/qsl/worldgen/biome/mixin/modification/DynamicRegistryManagerImplMixin.java [hunk 10]
```diff
@@ -0,0 +1,43 @@
...
+ * This Mixin allows us to keep backup copies of biomes for
+ * {@link BiomeModificationImpl} on a per-DynamicRegistryManager basis.
+ */
+@Mixin(DynamicRegistryManager.ImmutableRegistryManager.class)
+public class DynamicRegistryManagerImplMixin implements BiomeModificationMarker {
+	@Unique
+	private boolean modified;
+
+	@Override
+	public void quilt_markModified() {
```
Comment (LambdAurora) [inline library/worldgen/biome/src/main/java/org/quiltmc/qsl/worldgen/biome/mixin/modification/DynamicRegistryManagerImplMixin.java]:
```suggestion
	public void quilt$markModified() {
```

File: library/worldgen/biome/src/main/java/org/quiltmc/qsl/worldgen/biome/impl/modification/BiomeModificationImpl.java [hunk 11]
```diff
@@ -0,0 +1,232 @@
...
+
+	@SuppressWarnings("ConstantConditions")
+	public void finalizeWorldGen(DynamicRegistryManager impl, LevelProperties levelProperties) {
+		Stopwatch sw = Stopwatch.createStarted();
+
+		// Now that we apply biome modifications inside the MinecraftServer constructor, we should only ever do
+		// this once for a dynamic registry manager. Marking the dynamic registry manager as modified ensures a crash
+		// if the precondition is violated.
+		BiomeModificationMarker modificationTracker = (BiomeModificationMarker) impl;
+		modificationTracker.quilt_markModified();
```
Comment (LambdAurora) [inline library/worldgen/biome/src/main/java/org/quiltmc/qsl/worldgen/biome/impl/modification/BiomeModificationImpl.java]:
```suggestion
		modificationTracker.quilt$markModified();
```

File: library/worldgen/biome/src/main/java/org/quiltmc/qsl/worldgen/biome/impl/modification/BiomeModificationImpl.java [hunk 12]
```diff
@@ -0,0 +1,232 @@
...
+	}
+
+	@SuppressWarnings("ConstantConditions")
+	public void finalizeWorldGen(DynamicRegistryManager impl, LevelProperties levelProperties) {
+		Stopwatch sw = Stopwatch.createStarted();
+
+		// Now that we apply biome modifications inside the MinecraftServer constructor, we should only ever do
+		// this once for a dynamic registry manager. Marking the dynamic registry manager as modified ensures a crash
+		// if the precondition is violated.
+		BiomeModificationMarker modificationTracker = (BiomeModificationMarker) impl;
```
Comment (LambdAurora) [inline library/worldgen/biome/src/main/java/org/quiltmc/qsl/worldgen/biome/impl/modification/BiomeModificationImpl.java]:
```suggestion
		var modificationTracker = (BiomeModificationMarker) impl;
```

**Review by LambdAurora — APPROVED** [review]
Looking at this, and I realize there's quite an annoying flaw, but it's not blocking this PR as it can be introduced later neatly thanks to the context objects.

Here's an use-case that as-is it's a bit hacky to fulfill, not sure it's blocking for this API since the current way it works could allow us to modify the contexts later.

I was wondering about adding new worldgen features to a mod, though since worldgen is not the main focus of the mod it might add stuff players don't want, so I need a way to disable it, and per-world. The idea is simple: have a custom resource reloader that pass before Vanilla (#112), meaning a data-pack could configure this for a world, then depending on that either add or not the features.
The issue here is it requires to pass values, and currently threre's no clean way to pass those, we might want to consider looking into that for the context objects in the future.

Now, time for nit-picks:
I **really** would appreciate the use of the `this` keyword when a class member is called.
Even if it can sound redundant, it really helps when reviewing, especially in browser where you don't have a fancy IDE telling you where stuff comes from.

Aside from the other comments, sounds fine!
Please make sure to test entirely this PR and send a comment saying it was.

**Review by zOnlyKroks — APPROVED** [review]

**Comment by zOnlyKroks** [comment]
Pr is tested for server and client, both work!

**Comment by TheGlitch76** [comment]
I know status checks pass besides the testmod's QMJ, so I'm going to go ahead and put this into FCP.
@peaceheis please fix it before next Thursday!

**Pull request merged**